chainerx_cc/chainerx/routines/reduction.cc

```diff
@@ -98,4 +98,56 @@ Array LogSoftmax(const Array& x, const OptionalAxes& axis) {
     return x_cast - LogSumExp(x_cast, axis.has_value() ? axis : OptionalAxes{1}, true);
 }
 
+Array Cumsum(const Array& a, absl::optional<int8_t> axis) {
+    int8_t axis_norm;
+    Array a_reshaped{};
+    if (axis.has_value()) {
+        axis_norm = internal::NormalizeAxis(*axis, a.ndim());
+        a_reshaped = Copy(a);
+    } else {
+        axis_norm = 0;
+        // TODO(imanishi): Fix after chainerx::Ravel is supported.
+        a_reshaped = a.Reshape(Shape{a.GetTotalSize()});
+    }
+
+    Shape out_shape = a_reshaped.shape();
+    Array out = Empty(out_shape, a_reshaped.dtype(), a_reshaped.device());
+    // Decide the output dtype for integral input dtype.
+    Dtype out_dtype{};
+    switch (GetKind(a_reshaped.dtype())) {
+        case DtypeKind::kBool:
+        case DtypeKind::kInt:  // fallthrough
+            out_dtype = Dtype::kInt64;
+            break;
+        case DtypeKind::kUInt:
+            out_dtype = Dtype::kInt64;  // TODO(niboshi): This should be kUInt64
+            break;
+        default:
+            out_dtype = a_reshaped.dtype();
+    }
+
+    const Array& out_cast = out.AsType(out_dtype);
+    const Array& a_cast = a_reshaped.dtype() != out_cast.dtype() ? a_reshaped.AsType(out_cast.dtype()) : a_reshaped;
```

Comment:
Please move this cast to the device implementation.

Reply:
Done.